File: backend/src/xfd_django/xfd_api/schema_models/metrics.py
"""Schemas to support scan task endpoints."""

# Standard Python Libraries
from datetime import datetime
from typing import List, Optional

# Third-Party Libraries
from pydantic import BaseModel


class OrgCountByStatus(BaseModel):
    """HTTP status and distinct org count."""

    http_status: int
    org_count: int


class ScanOrgCountByStatus(BaseModel):
    """Scan metadata and success counts per status."""

    id: str
    created_at: datetime
    updated_at: datetime
    name: str
    frequency: int
    last_run: Optional[datetime]
    total_orgs: int
    org_counts_by_status: List[OrgCountByStatus]


class ListScansOrgCountByStatusResponse(BaseModel):
    """Top-level response model for API."""

    scans: List[ScanOrgCountByStatus]
    metrics_window_days: int


class DailyCount(BaseModel):
    """Count for a specific date."""

    date: str
    count: int


class DailyStatusCount(BaseModel):
    """Daily Counts for a specific http_status."""

    http_status: int
    daily_counts: List[DailyCount]


class GetScanDailyStatusCountsResponse(BaseModel):
    """Top-level response model for daily status counts."""

    id: str
    created_at: datetime
    updated_at: datetime
    name: str
    frequency: int
    last_run: Optional[datetime]
    total_orgs: int
    daily_status_counts: List[DailyStatusCount]
    metrics_window_days: int

File: frontend/src/types/metrics.ts
export interface OrgCountByStatus {
  http_status: number;
  org_count: number;
}

export interface ScanSummary {
  total_orgs: number;
  id: string;
  name: string;
  org_counts_by_status: OrgCountByStatus[];
}

export interface ScanSummaries {
  scans: ScanSummary[];
  metrics_window_days: number;
}

export interface DailyCount {
  date: string;
  count: number;
}

export interface DailyStatusCount {
  http_status: number;
  daily_counts: DailyCount[];
}

export interface ScanDetails {
  id: string;
  created_at: string;
  updated_at: string;
  name: string;
  frequency: number;
  last_run: string | null;
  total_orgs: number;
  daily_status_counts: DailyStatusCount[];
  metrics_window_days: number;
}
